NLL diagnostics: revise `fn check_access_permissions`

pnkfelix opened the pull request:
NLL: revise `fn check_access_permissions` so that its (still branchy) shares more code paths between the different cases, and also provide more diagnostics in more cases (though the added diagnostics still do not always meet the quality bar established by AST-borrowck)

----

Transcribing "checklist" suggested by Niko, except I am rendering it as a table to make it clear that I do not regard every item in the list to be a "must have" for landing this PR.

goal | does this PR do it?
-----|------------------------------
no suggestions for `ref mut` |  yes
suggestions for direct local assignment (`{ let x = 3; x = 4; }`) | yes (see commits at end)
suggestions for direct field assignment (`{ let x = (3, 4); x.0 = 5; }` | yes (see commits at end)
suggestions for upvars (`let x = 3; let c = \|\| { &mut x; }`) | yes

Note that I added support for a couple of rows via changes that are not strictly part of `fn check_access_permissions`. If desired I can remove those commits from this PR and leave them for a later PR. 

Fix #51031 
Fix #51032 
(bug #51191 needs a little more investigation before closing.)
Fix #51578

rust-highfive commented:
r? @cramertj

(rust_highfive has picked a reviewer for you, use r? to override)

nikomatsakis commented on src/librustc_mir/borrow_check/mod.rs:
```diff
@@ -1808,15 +1724,162 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
+            }
+        }
+
+        if note_the_place_err && place != the_place_err {
```
I personally would remove this message. If we can identify such a place, we should suggest how to change it to mutable. Otherwise, I suspect it will confuse users as much as anything.

pnkfelix replied:
My plan is indeed to remove this message once we’ve added all the diagnostic cases of interest

nikomatsakis commented on src/librustc_mir/borrow_check/mod.rs:
```diff
@@ -1808,15 +1724,162 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
+                // ... but it doesn't make sense to suggest it on
+                // variables that are already `ref x` or `ref mut x`
+                // (such variables are simply not mutable).
+                if let Some(ty::BindingMode::BindByValue(_)) =
```
In such cases, often the user should have written `&mut *foo` instead of `&mut foo` — maybe we can think how to suggest that. (I've seen this before...)

nikomatsakis commented on src/librustc_mir/borrow_check/mod.rs:
```diff
@@ -1808,15 +1724,162 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
+                // (such variables are simply not mutable).
+                if let Some(ty::BindingMode::BindByValue(_)) =
+                    self.mir.local_decls[*local].is_user_variable {
+                        true
```
What about upvars? We should test this example:

```
let x = (3, );
let c = || x.0 += 1;
```

and see that we issue a suggestion here to make it `let mut x`

pnkfelix replied:
Yeah, if #51031 does not currently include any examples of upvars, then we should make sure that those cases are being tested... I know I haven't done anything special to account for them ... **yet!**

nikomatsakis commented:
This check-list is from #51260 — some of these cases I think you handle, but not all...

- [ ] no suggestions for `ref mut`
- [ ] check on suggestions for other cases (`{ let x = 3; x = 4; }`)
  - oh, that seems to generate a slightly different message...maybe that's ok?
  - `{ let x = (3, 4); x.0 = 5; }` generates a sort-of strange message maybe? ([playground](https://play.rust-lang.org/?gist=1ed980ff940315c7d81f801cc5a28674&version=nightly&mode=debug) — compare NLL and not NLL)
- [ ] suggestions for upvars (`let x = 3; let c = || { &mut x; }`) ([playground](https://play.rust-lang.org/?gist=b7e4361660f79983ba4f9b2306439041&version=undefined&mode=undefined))

pnkfelix commented:
I want to preserve this comment from @nikomatsakis somewhere that won't become outdated by updated to the commit series:

https://github.com/pnkfelix/rust/blob/79e9c3d3fa5ad3b22141968c3a88e44475df75f3/src/librustc_mir/borrow_check/mod.rs#L1795-L1798

> In such cases, often the user should have written `&mut *foo` instead of `&mut foo` — maybe we can think how to suggest that. (I've seen this before...)

pnkfelix commented:
Current state is closer but still WIP.

In particular:
 1. I want to try to expand the scope of the PR just a little bit further to cover more diagnostics (mostly so other people know that they should focus on other NLL diagnostic issues not related to this function). I have tried to write notes in all such potential target issues.
 2. The changes to the NLL diagnostics are not quite complete yet and might be acceptable with some further tweaks. For example, in `borrowck-borrow-overloaded-deref-mut.rs`, AST-borrowck preserves the explicitly-written lifetime in its suggested output, while NLL leaves it out (which is almost certainly going to them yield erroneous code if the developer follows the suggestion literally).

pnkfelix commented:
(Oh and clearly I will need to rebase too. Which is just as well because I also want to refactor the commit series before actually submitting them for formal review.)

pnkfelix commented:
So I have some prototype commits that attempt to thread information about explicitly provided types, when present, down into the `mir::LocalDecl`.

This does seem to be workable in terms of improving the diagnostics for NLL.

But something is wrong with my approach because I think something I did to the compiler data structure has caused the dep-graph tests in `compile-fail` to start failing.

I'm going to push the commits in case @nikomatsakis has a chance to peek at them. But I also assume that he won't really have any time before I get a chance to start fresh on this tomorrow.

pnkfelix commented:
Ah i just realized: I bet the problem in the commits I pushed is that I added NodeId’s to the state of the HIR and MIR, and those won’t be stable in the face of other changes to the crate, which messes up DepInfo (right?)

Luckily I don’t *need* the NodeId; I just had it there in case I wanted to try to use it for something.

(Still, I am curious about what I should be inserting here to capture this state. Or maybe I should just deliberately omit the NodeId from the implementations of things like hash_stable?

pnkfelix commented:
Okay I've figured out how to use `ClearCrossCrate` to hide the threaded `HirId`. Things seem to work.

pnkfelix commented:
I think I'm going to make one last push to see if I can thread down the information about the initializer expression span in something like `let x = <expr>;` (we use that to do things like suggest "maybe use `&mut <expr>`)

Once that experiment is done (whether it works or not), I'll switch to focusing on cleaning up this PR and getting it ready for review.

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/391378360) ([raw log](https://api.travis-ci.org/v3/job/391378360/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:09:02]    Compiling proc_macro v0.0.0 (file:///checkout/src/libproc_macro)
[00:09:21]    Compiling syntax_ext v0.0.0 (file:///checkout/src/libsyntax_ext)
[00:15:39]    Compiling rustc_mir v0.0.0 (file:///checkout/src/librustc_mir)
[00:15:39]    Compiling rustc_typeck v0.0.0 (file:///checkout/src/librustc_typeck)
[00:15:41] error[E0425]: cannot find value `syntactic_scope` in this scope
[00:15:41]     --> librustc_mir/build/matches/mod.rs:1160:17
[00:15:41]      |
[00:15:41] 1160 |                 syntactic_scope,
[00:15:41] 
[00:15:41] 
[00:15:45] error[E0560]: struct `rustc::mir::LocalDecl<'tcx>` has no field named `syntactic_scope`
[00:15:45]     --> librustc_mir/build/matches/mod.rs:1160:17
[00:15:45]      |
[00:15:45] 1160 |                 syntactic_scope,
[00:15:45]      |                 ^^^^^^^^^^^^^^^ `rustc::mir::LocalDecl<'tcx>` does not have this field
[00:15:45]      |
[00:15:45]      = note: available fields are: `mutability`, `is_user_variable`, `internal`, `ty`, `name` ... and 2 others
[00:15:52] error: aborting due to 2 previous errors
[00:15:52] 
[00:15:52] Some errors occurred: E0425, E0560.
[00:15:52] For more information about an error, try `rustc --explain E0425`.
[00:15:52] For more information about an error, try `rustc --explain E0425`.
[00:15:52] error: Could not compile `rustc_mir`.
[00:15:52] 
[00:15:52] Caused by:
[00:15:52]   process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc --crate-name rustc_mir librustc_mir/lib.rs --color always --error-format json --crate-type dylib --emit=dep-info,link -C prefer-dynamic -C opt-level=2 -C metadata=1f35ea53eb60b9cb -C extra-filename=-1f35ea53eb60b9cb --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/release/deps --extern bitflags=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libbitflags-99b4534960f92449.rlib --extern byteorder=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libbyteorder-78ebc049d4f53f46.rlib --extern serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libserialize-5d0a8a65bb9fe29f.so --extern serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libserialize-5d0a8a65bb9fe29f.rlib --extern log_settings=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/liblog_settings-f16e9eee09cda644.rlib --extern arena=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libarena-ccde2368d50449de.so --extern log=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/liblog-5073f1296cd24b67.rlib --extern rustc_data_structures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_data_structures-a7df2fc298b4fb06.so --extern rustc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc-01e98699f9876faa.so --extern graphviz=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libgraphviz-be994de8b3050feb.so --extern syntax_pos=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libsyntax_pos-670f6bac60d99b34.so --extern rustc_errors=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_errors-c9d6678b1c0f0b46.so --extern polonius_engine=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libpolonius_engine-6e0be5cf77966185.rlib --extern rustc_apfloat=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_apfloat-642a7efe79e24835.rlib --extern syntax=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libsyntax-6a2f0731783c2bd3.so --extern either=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libeither-11fcd35bf55f48b3.rlib --extern rustc_target=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_target-58741ed9de9aae4f.so -L native=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/build/backtrace-sys-88673787176f9d86/out -L native=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/build/miniz-sys-62d80197b9ec531e/out` (exit code: 101)
rustlib/x86_64-unknown-linux-gnu
315008 ./src/llvm
249872 ./obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib
241180 ./src/llvm-emscripten
---
147592 ./.git/modules/src
137992 ./obj/build/bootstrap/debug/incremental
132184 ./obj/build/x86_64-unknown-linux-gnu/stage0-rustc
123420 ./obj/build/bootstrap/debug/incremental/bootstrap-1r3bppl29tbrj
123416 ./obj/build/bootstrap/debug/incremental/bootstrap-1r3bppl29tbrj/s-f1x8fd6lok-6wniw6-bfrkfwrvh0tv
89804 ./src/llvm/test/CodeGen
81432 ./obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu
81428 ./obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release
72472 ./.git/modules/src/tools
---
travis_time:end:0589127d:start=1528825746652078033,finish=1528825746658101010,duration=6022977
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:12589860
$ head -30 ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
head: cannot open ‘./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers’ for reading: No such file or directory
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:0bb7026b
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/391387552) ([raw log](https://api.travis-ci.org/v3/job/391387552/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:start:test_ui_nll
Check compiletest suite=ui mode=ui compare_mode=nll (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[00:50:29] 
[00:50:29] running 1502 tests
[00:50:34] ...............................................F............................................i.......
[00:50:40] ......................................................i................F............................
[00:50:48] ....................................................................................................
[00:50:52] ....................................................................................................
[00:50:55] ....................................................................................................
[00:51:00] ....................................................................................................
[00:51:00] ....................................................................................................
[00:51:05] ....................................................................................................
[00:51:10] ....................................................................................................
[00:51:16] ....................................................................................................
[00:51:21] ....i..............................................................................i................
[00:51:27] ....................................................................................................
[00:51:32] ....................................................................................................
[00:51:37] ....................................................................................................
[00:51:43] .................i.....................F............................................................
[00:51:44] failures:
[00:51:44] 
[00:51:44] ---- [ui (nll)] ui/borrowck/mut-borrow-of-mut-ref.rs stdout ----
[00:51:44] diff of stderr:
[00:51:44] diff of stderr:
[00:51:44] 
[00:51:44] 1 error[E0596]: cannot borrow immutable item `b` as mutable
[00:51:44] 3    |
[00:51:44] 3    |
[00:51:44] + LL | fn f(b: &mut i32) {
[00:51:44] +    |      - help: consider changing this to be mutable: `mut b`
[00:51:44] 4 LL |     g(&mut b) //~ ERROR cannot borrow
[00:51:44] 5    |       ^^^^^^ cannot borrow as mutable
[00:51:44] 
[00:51:44] 
[00:51:44] The actual stderr differed from the expected stderr.
[00:51:44] The actual stderr differed from the expected stderr.
[00:51:44] Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/borrowck/mut-borrow-of-mut-ref.nll/mut-borrow-of-mut-ref.nll.stderr
[00:51:44] To update references, rerun the tests and pass the `--bless` flag
[00:51:44] To only update this specific test, also pass `--test-args borrowck/mut-borrow-of-mut-ref.rs`
[00:51:44] error: 1 errors occurred comparing output.
[00:51:44] status: exit code: 101
[00:51:44] status: exit code: 101
[00:51:44] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/borrowck/mut-borrow-of-mut-ref.rs" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/borrowck/mut-borrow-of-mut-ref.nll/a" "-Zborrowck=mir" "-Ztwo-phase-borrows" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linuxd","expansion":null}],"children":[],"rendered":null}],"rendered":"error[E0596]: cannot borrow immutable item `b` as mutable\n  --> /checkout/src/test/ui/borrowck/mut-borrow-of-mut-ref.rs:18:7\n   |\nLL | fn f(b: &mut i32) {\n   |      - help: consider changing this to be mutable: `mut b`\nLL |     g(&mut b) //~ ERROR cannot borrow\n   |       ^^^^^^ cannot borrow as mutable\n\n"}
[00:51:44] {"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
[00:51:44] {"message":"For more information about this error, try `rustc --explain E0596`.","code":null,"level":"","spans":[],"children":[],"rendered":"For more information about this error, try `rustc --explain E0596`.\n"}
[00:51:44] ------------------------------------------
[00:51:44] 
[00:51:44] thread '[ui (nll)] ui/borrowck/mut-borrow-of-mut-ref.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3139:9
[00:51:44] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:51:44] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:51:44] 
[00:51:44] ---- [ui (nll)] ui/did_you_mean/issue-31424.rs stdout ----
[00:51:44] diff of stderr:
[00:51:44] 
[00:51:44] 8   --> $DIR/issue-31424.rs:23:9
[00:51:44] 9    |
[00:51:44] 10 LL |     fn bar(self: &mut Self) {
[00:51:44] -    |            --------------- help: consider changing this to be mutable: `mut self`
[00:51:44] +    |            ---- help: consider changing this to be mutable: `mut self`
[00:51:44] 12 LL |         (&mut self).bar(); //~ ERROR cannot borrow
[00:51:44] 13    |         ^^^^^^^^^^^ cannot borrow as mutable
[00:51:44] 
[00:51:44] 
[00:51:44] The actual stderr differed from the expected stderr.
[00:51:44] The actual stderr differed from the expected stderr.
[00:51:44] Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/did_you_mean/issue-31424.nll/issue-31424.nll.stderr
[00:51:44] To update references, rerun the tests and pass the `--bless` flag
[00:51:44] To only update this specific test, also pass `--test-args did_you_mean/issue-31424.rs`
[00:51:44] error: 1 errors occurred comparing output.
[00:51:44] status: exit code: 101
[00:51:44] status: exit code: 101
[00:51:44] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/did_you_mean/issue-31424.rs" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/did_you_mean/issue-31424.nll/a" "-Zborrowck=mir" "-Ztwo-phase-borrows" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/did_you_mean/issue-31424.nll/auxiliary" "-A" "unused"
[00:51:44] ------------------------------------------
[00:51:44] 
[00:51:44] ------------------------------------------
[00:51:44] stderr:
[00:51:44] stderr:
[00:51:44] ------------------------------------------
[00:51:44] {"message":"cannot borrow immutable item `self` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so light_end":20}],"label":"cannot borrow as mutable","suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"consider changing this to be mutable","code":null,"level":"help","spans":[{"file_name":"/checkout/src/test/ui/did_you_mean/issue-31424.rs","byte_start":758,"byte_end":762,"line_start":22,"line_end":22,"column_start":12,"column_end":16,"is_primary":true,"text":[{"text":"    fn bar(self: &mut Self) {","highlight_start":12,"highlight_end":16}],"label":null,"suggested_replacement":"mut self","suggestion_applicability":"Unspecified","expansion":null}],"children":[],"rendered":null}],"rendered":"error[E0596]: cannot borrow immutable item `self` as mutable\n  --> /checkout/src/test/ui/did_you_mean/issue-31424.rs:23:9\n   |\nLL |     fn bar(self: &mut Self) {\n   |            ---- help: consider changing this to be mutable: `mut self`\nLL |         (&mut self).bar(); //~ ERROR cannot borrow\n   |         ^^^^^^^^^^^ cannot borrow as mutable\n\n"}
[00:51:44] {"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 2 previous errors\n\n"}
[00:51:44] {"message":"For more information about this error, try `rustc --explain E0596`.","code":null,"level":"","spans":[],"children":[],"rendered":"For more information about this error, try `rustc --explain E0596`.\n"}
[00:51:44] ------------------------------------------
[00:51:44] 
[00:51:44] thread '[ui (nll)] ui/did_you_mean/issue-31424.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3139:9
[00:51:44] 
[00:51:44] 
[00:51:44] ---- [ui (nll)] ui/trivial-bounds-inconsistent-copy-reborrow.rs stdout ----
[00:51:44] diff of stderr:
[00:51:44] 
[00:51:44] 1 error[E0596]: cannot borrow immutable item `**t` as mutable
[00:51:44] 3    |
[00:51:44] 3    |
[00:51:44] + LL | fn reborrow_mut<'a>(t: &'a &'a mut i32) -> &'a mut i32 where &'a mut i32: Copy {
[00:51:44] +    |                        --------------- help: consider changing this to be a mutable reference: `&mut &mut i32`
[00:51:44] 4 LL |     *t //~ ERROR
[00:51:44] 5    |     ^^ cannot borrow as mutable
[00:51:44] -    |
[00:51:44] -    = note: the value which is causing this path not to be mutable is...: `*t`
[00:51:44] 8 
[00:51:44] 9 error[E0596]: cannot borrow immutable item `**t` as mutable
[00:51:44] 
[00:51:44] 11    |
[00:51:44] 11    |
[00:51:44] + LL | fn copy_reborrow_mut<'a>(t: &'a &'a mut i32) -> &'a mut i32 where &'a mut i32: Copy {
[00:51:44] +    |                             --------------- help: consider changing this to be a mutable reference: `&mut &mut i32`
[00:51:44] 12 LL |     {*t} //~ ERROR
[00:51:44] 13    |      ^^ cannot borrow as mutable
[00:51:44] -    |
[00:51:44] -    = note: the value which is causing this path not to be mutable is...: `*t`
[00:51:44] 17 error: aborting due to 2 previous errors
[00:51:44] 18 
[00:51:44] 
[00:51:44] 
[00:51:44] 
[00:51:44] The actual stderr differed from the expected stderr.
[00:51:44] Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/trivial-bounds-inconsistent-copy-reborrow.nll/trivial-bounds-inconsistent-copy-reborrow.nll.stderr
[00:51:44] To update references, rerun the tests and pass the `--bless` flag
[00:51:44] To only update this specific test, also pass `--test-args trivial-bounds-inconsistent-copy-reborrow.rs`
[00:51:44] error: 1 errors occurred comparing output.
[00:51:44] status: exit code: 101
[00:51:44] status: exit code: 101
[00:51:44] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/trivial-bounds-inconsistent-copy-reborrow.rs" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/trivial-bounds-inconsistent-copy-reborrow.nll/a" "-Zborrowck=mir" "-Ztwo-phase-borrows" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/trivial-bounds-inconsistent-copy-reborrow.nll/auxiliary" "-A" "unused"
[00:51:44] ------------------------------------------
[00:51:44] 
[00:51:44] ------------------------------------------
[00:51:44] stderr:
[00:51:44] stderr:
[00:51:44] ------------------------------------------
[00:51:44] {"message":"cannot borrow immutable item `**t` as mutable","code":{"code":"E0596","explanation":"\nThis error occurs because you tried to mutably borrow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` murow a non-mutable variable.\n\nExample of erroneous code:\n\n```compile_fail,E0596\nlet x = 1;\nlet y = &mut x; // error: cannot borrow mutably\n```\n\nIn here, `x` isn't mutable, so when we try to mutably borrow it in `y`, it\nfails. To fix this error, you need to make `x` mutable:\n\n```\nlet mut x = 1;\nlet y = &mut x; // ok!\n```\n"},"level":"error","spans":[{"file_name":"/checkout/src/test/ui/trivial-bounds-inconsistent-copy-reborrow.rs","byte_start":777,"byte_end":779,"line_start":20,"line_end":20,"column_start":6,"column_end":8,"is_primary":true,"text":[{"text":"    {*t} //~ ERROR","highlight_start":6,"highlight_end":8}],"label":"cannot borrow as mutable","suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"consider changing this to be a mutable reference","code":null,"level":"help","spans":[{"file_name":"/checkout/src/test/ui/trivial-bounds-inconsistent-copy-reborrow.rs","byte_start":714,"byte_end":729,"line_start":19,"line_end":19,"column_start":29,"column_end":44,"is_primary":true,"text":[{"text":"fn copy_reborrow_mut<'a>(t: &'a &'a mut i32) -> &'a mut i32 where &'a mut i32: Copy {","highlight_start":29,"highlight_end":44}],"label":null,"suggested_replacement":"&mut &mut i32","suggestion_applicability":"Unspecified","expansion":null}],"children":[],"rendered":null}],"rendered":"error[E0596]: cannot borrow immutable item `**t` as mutable\n  --> /checkout/src/test/ui/trivial-bounds-inconsistent-copy-reborrow.rs:20:6\n   |\nLL | fn copy_reborrow_mut<'a>(t: &'a &'a mut i32) -> &'a mut i32 where &'a mut i32: Copy {\n   |                             --------------- help: consider changing this to be a mutable reference: `&mut &mut i32`\nLL |     {*t} //~ ERROR\n   |      ^^ cannot borrow as mutable\n\n"}
[00:51:44] {"message":"aborting due to 2 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 2 previous errors\n\n"}
[00:51:44] {"message":"For more information about this error, try `rustc --explain E0596`.","code":null,"level":"","spans":[],"children":[],"rendered":"For more information about this error, try `rustc --explain E0596`.\n"}
[00:51:44] ------------------------------------------
[00:51:44] 
[00:51:44] thread '[ui (nll)] ui/trivial-bounds-inconsistent-copy-reborrow.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3139:9
[00:51:44] 
---
[00:51:44] 
[00:51:44] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:498:22
[00:51:44] 
[00:51:44] 
[00:51:44] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rusd https://google.com | grep ^Date: | sed 's/Date: //g' || true)
travis_fold:start:after_failure.1
travis_time:start:00b4d0ab
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/391510086) ([raw log](https://api.travis-ci.org/v3/job/391510086/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:start:test_incremental
Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:00:07] 
[01:00:07] running 89 tests
highlight_end":8},{"text":"}","highlight_start":1,"highlight_end":2}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"error: `MirOptimized(add_else_branch)` should be clean but is not\n  --> /checkout/src/test/incremental/hashes/if_expressions.rs:110:1\n   |\nLL | / pub fn add_else_branch(x: bool) -> u32 {\nLL | |     let mut ret = 1;\nLL | |\nLL | |     if x {\n...  |\nLL | |     ret\nLL | | }\n   | |_^\n\n"}
[01:00:29] {"message":"`MirValidated(add_else_branch)` should be clean but is not","code":null,"level":"error","spans":[{"file_name":"/checkout/src/test/incremental/hashes/if_expressions.rs","byte_start":2526,"byte_end":2646,"line_start":110,"line_end":119,"column_start":1,"column_end":2,"is_primary":true,"text":[{"text":"pub fn add_else_branch(x: bool) -> u32 {","highlight_start":1,"highlight_end":41},{"text":"    let mut ret = 1;","highlight_start":1,"highlight_end":21},{"text":"","highlight_start":1,"highlight_end":1},{"text":"    if x {","highlight_start":1,"highlight_end":11},{"text":"        ret = 2;","highlight_start":1,"highlight_end":17},{"text":"    } else {","highlight_start":1,"highlight_end":13},{"text":"    }","highlight_start":1,"highlight_end":6},{"text":"","highlight_start":1,"highlight_end":1},{"text":"    ret","highlight_start":1,"highlight_end":8},{"text":"}","highlight_start":1,"highlight_end":2}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"error: `MirValidated(add_else_branch)` should be clean but is not\n  --> /checkout/src/test/incremental/hashes/if_exf_let)` should be clean but is not","code":null,"level":"error","spans":[{"file_name":"/checkout/src/test/incremental/hashes/if_expressions.rs","byte_start":4393,"byte_end":4541,"line_start":207,"line_end":216,"column_start":1,"column_end":2,"is_primary":true,"text":[{"text":"pub fn add_else_branch_if_let(x: Option<u32>) -> u32 {","highlight_start":1,"highlight_end":55},{"text":"    let mut ret = 1;","highlight_start":1,"highlight_end":21},{"text":"","highlight_start":1,"highlight_end":1},{"text":"    if let Some(x) = x {","highlight_start":1,"highlight_end":25},{"text":"        ret = x;","highlight_start":1,"highlight_end":17},{"text":"    } else {","highlight_start":1,"highlight_end":13},{"text":"    }","highlight_start":1,"highlight_end":6},{"text":"","highlight_start":1,"highlight_end":1},{"text":"    ret","highlight_start":1,"highlight_end":8},{"text":"}","highlight_start":1,"highlight_end":2}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"error: `MirOptimized(add_else_branch_if_let)` should be clean but is not\n  --> /checkout/src/test/incremental/hashes/if_expressions.rs:207:1\n   |\nLL | / pub fn add_else_branch_if_let(x: Option<u32>) -> u32 {\nLL | |     let mut ret = 1;\nLL | |\nLL | |     if let Some(x) = x {\n...  |\nLL | |     ret\nLL | | }\n   | |_^\n\n"}
[01:00:29] {"message":"aborting due to 4 previous errors","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to 4 previous errors\n\n"}
[01:00:29] ------------------------------------------
[01:00:29] 
[01:00:29] thread '[incremental] incremental/hashes/if_expressions.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3139:9
[01:00:29] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:00:29] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:00:29] 
[01:00:29] ---- [incremental] incremental/hashes/trait_defs.rs stdout ----
[01:00:29] 
[01:00:29] error in revision `cfail2`: test compilation failed although it shouldn't!
[01:00:29] status: exit code: 101
[01:00:29] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/incremental/hashes/trait_defs.rs" "--target=x86_64-unknown-linux-gnu" "--cfg" "cfail2" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/trait_defs/trait_defs.inc" "-Z" "incremental-verify-ich" "-Z" "incremental-queries" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/trait_defs/a" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Z" "query-dep-graph" "-Zincremental-ignore-spans" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/trait_defs/auxiliary"
[01:00:29] ------------------------------------------
[01:00:29] 
[01:00:29] ------------------------------------------
[01:00:29] stderr:
[01:00:29] stderr:
[01:00:29] ------------------------------------------
[01:00:29] {"message":"`Hir(TraitChangeModeSelfOwnToMut::method)` should be clean but is not","code":null,"level":"error","spans":[{"file_name":"/checkout/src/test/incremental/hashes/trait_defs.rs","byte_start":7313,"byte_end":7335,"line_start":277,"line_end":277,6_64-unknown-linux-gnu/stage1-rustc/release/build
34840 ./obj/build/x86_64-unknown-linux-gnu/test/run-pass
34588 ./obj/build/x86_64-unknown-linux-gnu/native/jemalloc/lib
34372 ./obj/build/x86_64-unknown-linux-gnu/doc/core/arch
33884 ./src/llvm-emscripten/lib/Target
---
travis_time:end:26529e85:start=1528849055005141390,finish=1528849055012084695,duration=6943305
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:132a6624
$ head -30 ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
head: cannot open ‘./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers’ for reading: No such file or directory
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:0cf47ccd
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

pnkfelix commented:
I am ready to have this actually go through review.

I have not addressed all the points listed in [niko's above comment](https://github.com/rust-lang/rust/pull/51275#issuecomment-393968198), but the branch has gotten to the point where I think it is worth landing and then building further on top of it after it lands.

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/391551456) ([raw log](https://api.travis-ci.org/v3/job/391551456/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:start:test_incremental
Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:06:00] 
[01:06:00] running 89 tests
[01:06:23] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:498:22
[01:06:23] ..........................F.............F................................................
[01:06:23] 
[01:06:23] ---- [incremental] incremental/hashes/if_expressions.rs stdout ----
[01:06:23] 
[01:06:23] 
[01:06:23] error in revision `cfail2`: test compilation failed although it shouldn't!
[01:06:23] status: exit code: 101
[01:06:23] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/incremental/hashes/if_expressions.rs" "--target=x86_64-unknown-linux-gnu" "--cfg" "cfail2" "-C" "incremental=/checkout/objhighlight_end":8},{"text":"}","highlight_start":1,"highlight_end":2}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"error: `MirOptimized(add_else_branch)` should be clean but is not\n  --> /checkout/src/test/incremental/hashes/if_expressions.rs:110:1\n   |\nLL | / pub fn add_else_branch(x: bool) -> u32 {\nLL | |     let mut ret = 1;\nLL | |\nLL | |     if x {\n...  |\nLL | |     ret\nLL | | }\n   | |_^\n\n"}
[01:06:23] {"message":"`MirValidated(add_else_branch)` should be clean but is not","code":null,"level":"error","spans":[{"file_name":"/checkout/src/test/incremental/hashes/if_expressions.rs","byte_start":2526,"byte_end":2646,"line_start":110,"line_end":119,"column_start":1,"column_end":2,"is_primary":true,"text":[{"text":"pub fn add_else_branch(x: bool) -> u32 {","highlight_start":1,"highlight_end":41},{"text":"    let mut ret = 1;","highlight_start":1,"highlight_end":21},{"text":"","highlight_start":1,"highlight_end":1},{"text":"    if x {","highlight_start":1,"highlight_end":11},{"text":"        ret = 2;","highlight_start":1,"highlight_end":17},{"text":"    } else {","highlight_start":1,"highlight_end":13},{"text":"    }","highlight_start":1,"highlight_end":6},{"text":"","highlight_start":1,"highlight_end":1},{"text":"    ret","highlight_start":1,"highlight_end":8},{"text":"}","highlight_start":1,"highlight_end":2}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"error: `MirValidated(add_else_branch)` should be clean but is not\n  --> /checkout/src/test/incremental/hashes/if_exemental/hashes/if_expressions.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3139:9
[01:06:23] 
[01:06:23] ---- [incremental] incremental/hashes/trait_defs.rs stdout ----
[01:06:23] 
[01:06:23] 
[01:06:23] error in revision `cfail2`: test compilation failed although it shouldn't!
[01:06:23] status: exit code: 101
[01:06:23] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/incremental/hashes/trait_defs.rs" "--target=x86_64-unknown-linux-gnu" "--cfg" "cfail2" "-C" "incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/trait_defs/trait_defs.inc" "-Z" "incremental-verify-ich" "-Z" "incremental-queries" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/trait_defs/a" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Z" "query-dep-graph" "-Zincremental-ignore-spans" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/hashes/trait_defs/auxiliary"
[01:06:23] ------------------------------------------
[01:06:23] 
[01:06:23] ------------------------------------------
[01:06:23] stderr:
[01:06:23] stderr:
[01:06:23] ------------------------------------------
[01:06:23] {"message":"`Hir(TraitChangeModeSelfOwnToMut::method)` should be clean but is not","code":null,"level":"error","spans":[{"file_name":"/checkout/src/test/incremental/hashes/trait_defs.rs","byte_start":7313,"byte_end":7335,"line_start":277,"line_end":277,"column_start":5,"column_end":27,"is_primary":true,"text":[{"text":"    fn method(mut self) {}","highlight_start":5,"highlight_end":27}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":"error: `Hir(TraitChangeModeSelfOwnToMut::method)` should be clean but is not\n  --> /checkout/src/test/incremental/hashes/trait_defs.rs:277:5\n   |\nLL |     fn method(mut self) {}\n   |     ^^^^^^^^^^^^^^^^^^^^^^\n\n"}
[01:06:23] {"message":"aborting due to previous error","code":null,"level":"error","spans":[],"children":[],"rendered":"error: aborting due to previous error\n\n"}
[01:06:23] ------------------------------------------
[01:06:23] 
[01:06:23] thread '[incremental] incremental/hashes/trait_defs.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3139:9
[01:06:23] 
---
[01:06:23] test result: FAILED. 87 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out
[01:06:23] 
[01:06:23] 
[01:06:23] 
[01:06:23] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/incremental" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "incremental" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-3.9/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "3.9.1\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[01:06:23] 
[01:06:23] 
[01:06:23] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:06:23] Build completed unsuccessfully in 0:17:50
[01:06:23] Build completed unsuccessfully in 0:17:50
[01:06:23] make: *** [check] Error 1
[01:06:23] Makefile:58: recipe for target 'check' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:192cc01e
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:0b5384fd:start=1528859612727395794,finish=1528859612733696161,duration=6300367
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0a5be788
$ head -30 ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
head: cannot open ‘./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers’ for reading: No such file or directory
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:0d2c6fa2
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

pnkfelix commented:
Yikes I guess it was pretty dumb of me to be testing only my usual subset of the test suite (ui, compile-fail, run-pass, mir-opt) in this particular case where I am adding info to data structures utilized for incremental compilation....

pnkfelix commented:
Bisection indicates that the failure of the `incremental/hashes/trait_defs.rs` test is injected by the very first commit in this PR, 020d970897

I'm not quite sure what is wrong with what I have done there, though...

pnkfelix commented:
Ah I see what I did wrong.

I am attempting to thread-down what kind of "implicit self" parameter we have in the `hir::FnDecl`: is it `&self`, `&mut self`, `self`, or `mut self`.

The problem with that is that the `FnDecl` is meant to represent the externally visible *interface* to the function. From the external point of view, there is no difference between `self` and `mut self`; they are the same thing. So it is bad to have that distinction hanging onto the `FnDecl`.

(I suppose that from the external point of view there is also no different between an "implicit self" and an explicitly-typed version of the same parameter. Hmm. Maybe the cleanest thing would be to wrap this additional field on `FnDecl` in a `CleanCrossCrate`...)

pnkfelix commented:
Hmm actually on further inspection, it looks like the last round of "simplification" that I did actually could have gone further since I apparently no longer actually *use* the state that I threaded down through the HIR.

pnkfelix commented:
So it was easy to address the failure with `incremental/hashes/trait_defs.rs`, and I've pushed an updated branch accordingly.

But I cannot figure out what is going on with `incremental/hashes/if_expressions.rs`. Bisection indicates that the problem there is injected by 6a97d2ea88 but I cannot figure out how that change could have affected that test...

pnkfelix commented:
(my best guess is that I'm *still* having some issue associated with threading down the `hir::HirId` into the MIR, since I assume the representation of the `hir::HirId` is changing in response to the changes to the code, even though I thought that I encapsulated the threaded `hir::HirId` within a `ClearCrossCrate` ...)

pnkfelix commented:
Assuming this latest batch of commits passes travis, I am again ready to have this reviewed. I've added a table to the description on this PR that tries to describe (at least a subset of) the cases that are handled here...

nikomatsakis commented on src/test/ui/suggestions/fn-closure-mutable-capture.nll.stderr:
```diff
@@ -1,8 +1,8 @@
-error[E0594]: cannot assign to immutable item `x`
+error[E0594]: cannot assign to `x` which is behind a `&` reference
```
Hmm, maybe we should flag this issue as a follow-up work item — it seems a bit confusing, since the reference here is really introduced by the closure desugaring, right?

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit 032eee1 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 032eee1c1a7796fe73a76e2baed546c34105f2e1 -->

bors commented:
:hourglass: Testing commit 032eee1c1a7796fe73a76e2baed546c34105f2e1 with merge 0c04aad3ce41c2faffe79af5ad1b033b3fc1f624...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.7810)

kennytm commented:
Legit. Needs to update a UI test result.

```
---- [ui (nll)] ui\asm-out-assign-imm.rs stdout ----
diff of stderr:
1	error[E0384]: cannot assign twice to immutable variable `x`
-	  --> $DIR/asm-out-assign-imm.rs:31:9
+	  --> $DIR/asm-out-assign-imm.rs:33:9
3	   |
4	LL |     let x: isize;
5	   |         - consider changing this to `mut x`
```

pnkfelix commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 4684649 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 46846496ed6194d4cacf1fbd52b9c76b5c83b96c -->

bors commented:
:hourglass: Testing commit 46846496ed6194d4cacf1fbd52b9c76b5c83b96c with merge f28c7aef7fbff1909c2d0257186cd7a5f0c6aa4b...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.7820), [status-travis](https://travis-ci.org/rust-lang/rust/builds/394308439?utm_source=github_status&utm_medium=notification)
Approved by: nikomatsakis
Pushing f28c7aef7fbff1909c2d0257186cd7a5f0c6aa4b to master...
